app/models/concerns/order_filtering.rb

```diff
@@ -0,0 +1,96 @@
+# Filtering and priority logic for orders is extracted here to avoid cluttering the model class
+module OrderFiltering
+  extend ActiveSupport::Concern
+
+  def is_priority?
+    self.class.priority.exists?(self)
+  end
+
+  module ClassMethods
+    def priority
+      join_order_transports.where <<-SQL
+        (state = 'submitted' AND submitted_at::timestamptz <= timestamptz '#{one_day_ago}') OR
+        (state = 'processing' AND processed_at::timestamptz < timestamptz '#{last_6pm}') OR
+        (state = 'awaiting_dispatch' AND order_transports.scheduled_at::timestamptz < timestamptz '#{Time.now}') OR
+        (state = 'dispatching' AND dispatch_started_at::timestamptz < timestamptz '#{last_6pm}')
+      SQL
+    end
+
+    #
+    # Returns orders filtered using the following options :
+    #   - states[] A list of string matching the 'state' column of the order
+    #   - types[] A list of custom types, each defined by a combination or properties
+    #   - priority? Whether to only return orders deemed as 'high priority' or not
+    #
+    # Available types to filter on
+    #   - appointment
+    #   - online_order_pickup
+    #   - online_order_ggv
+    #   - shipment
+    #   - other
+    #
+    # To add a custom type, add an sql returning method formatted as {type_name}_sql
+    # e.g :
+    #   def shipment_sql
+    #     "detail_type = 'shipment'"
+    #   end
+    #
+    def filter(states: [], types: [], priority: false)
+      res = all
```

Comment:
I feel this will load all  records in memory and then it will fire where query. I feel we should remove this all. 

If this is class method then we can directly write `where("state IN (?)", states)` as it will be used by rails as `Order.where("state IN (?)", states)`

This way all records will not be loaded in memory first.

CC: @steveyken @Darshan-Chauhan @abulsayyad123

Reply:
@namrataukirde 
The point here is to do conditional chaining, we just need `res` to be a reference to the current relation on which we can apply the the 3 possible sql queries below.
basically it is possible that we call this method with no arguments, in which case it returns the unmodified relation

If `all` really does run the query then we should change it, maybe to an empty `where()` ?
The question becomes : How to get the current relation ?

Reply:
https://github.com/crossroads/api.goodcity/commit/6cbcac939f2cf451576df626d60a0f958c9c45bf